contracts/src/tests/test_rbac_timelock.cairo

```diff
@@ -0,0 +1,865 @@
+            ]
+        );
+
+    assert(!timelock.is_operation(id), 'not operation');
+    assert(timelock.get_timestamp(id) == 0, 'matches 0');
+}
+
+#[test]
+#[feature("safe_dispatcher")]
+fn test_execute_op_not_ready() {
+    let (target_address, _) = setup_mock_target();
+    let (min_delay, _, proposer, executor, _, _) = deploy_args();
+    let (_, timelock, safe_timelock) = setup_timelock();
+
+    let mock_time = 3;
+    let mock_ready_time = mock_time + min_delay.try_into().unwrap();
+    start_cheat_block_timestamp_global(mock_time);
+
+    let call_1 = Call {
+        target: target_address, selector: selector!("set_value"), data: array![0x56162].span()
+    };
+
+    let call_2 = Call {
+        target: target_address, selector: selector!("flip_toggle"), data: array![].span()
+    };
+
+    let calls = array![call_1, call_2].span();
+    let predecessor = 0;
+    let salt = 1;
+
+    start_cheat_caller_address_global(executor);
+
+    // test not scheduled
+    expect_operation_not_ready(safe_timelock.execute_batch(calls, predecessor, salt));
+
+    // test not ready
+
+    // first, schedule it
+    start_cheat_caller_address_global(proposer);
+    timelock.schedule_batch(calls, predecessor, salt, min_delay);
+
+    // then, execute at a block that is NOT ready
+    start_cheat_block_timestamp_global(mock_ready_time - 1);
+    start_cheat_caller_address_global(executor);
+    expect_operation_not_ready(safe_timelock.execute_batch(calls, predecessor, salt));
+
+    // execute, then test it can't be executed after it's done already
+    start_cheat_block_timestamp_global(mock_ready_time);
+    timelock.execute_batch(calls, predecessor, salt);
+
+    expect_operation_not_ready(safe_timelock.execute_batch(calls, predecessor, salt));
+}
+
+#[test]
+#[feature("safe_dispatcher")]
+fn test_execute_predecessor_invalid() {
+    let (min_delay, _, proposer, executor, _, _) = deploy_args();
+    let (_, timelock, safe_timelock) = setup_timelock();
+
+    let mock_time = 3;
+    let mock_ready_time = mock_time + min_delay.try_into().unwrap();
+    start_cheat_block_timestamp_global(mock_time);
+
+    let call = Call {
+        target: contract_address_const::<100>(),
+        selector: selector!("doesnt_exist"),
+        data: array![0x123].span()
+    };
+    let calls = array![call].span();
+    let predecessor = 4;
+    let salt = 1;
+
+    start_cheat_caller_address_global(proposer);
+    timelock.schedule_batch(calls, predecessor, salt, min_delay);
+
+    start_cheat_caller_address_global(executor);
+    start_cheat_block_timestamp_global(mock_ready_time);
+    let result = safe_timelock.execute_batch(calls, predecessor, salt);
+    match result {
+        Result::Ok(_) => panic!("expect 'rbact: missing dependency'"),
+        Result::Err(panic_data) => {
+            assert(*panic_data.at(0) == 'rbact: missing dependency', *panic_data.at(0));
+        }
+    }
+}
+
+// snforge does not treat contract invocation failures as a panic (yet)
```

Comment:
so executing this on a real environment would result in a panic? is the timelock contract able to tell which op failed?

Reply:
in a real environment you'd get a panic / revert. in general, the EVM version also does not explictly tell you which operation failed